hazelcast/src/test/java/com/hazelcast/collection/impl/queue/QueuePriorityWithDuplicateCheckTest.java

```diff
@@ -0,0 +1,224 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.collection.impl.queue;
+
+import com.hazelcast.collection.IQueue;
+import com.hazelcast.collection.impl.queue.model.PriorityElement;
+import com.hazelcast.collection.impl.queue.model.PriorityElementComparator;
+import com.hazelcast.config.Config;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.Logger;
+import com.hazelcast.map.IMap;
+import com.hazelcast.test.HazelcastParallelClassRunner;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.annotation.ParallelJVMTest;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import java.util.concurrent.ConcurrentSkipListSet;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelJVMTest.class})
+public class QueuePriorityWithDuplicateCheckTest extends HazelcastTestSupport {
+
+    private static final ILogger LOG = Logger.getLogger(QueuePriorityWithDuplicateCheckTest.class);
+    private PriorityElementTaskQueueImpl queue;
+
+    @Before
+    public void before() {
+        Config config = smallInstanceConfig();
+        String queueName = randomName();
+        String mapName = randomName();
+
+        config.getQueueConfig(queueName)
+              .setPriorityComparatorClassName("com.hazelcast.collection.impl.queue.model.PriorityElementComparator");
+        HazelcastInstance hz = createHazelcastInstance(config);
+        queue = new PriorityElementTaskQueueImpl(hz.getQueue(queueName), hz.getMap(mapName));
+    }
+
+    @Test
+    public void queue() {
+        PriorityElement element = new PriorityElement(false, 1);
+        queue.enqueue(element);
+        assertEquals(element, queue.dequeue());
+        assertNull(queue.dequeue());
+    }
+
+    @Test
+    public void queueDouble() {
+        PriorityElement element = new PriorityElement(false, 1);
+        queue.enqueue(element);
+        queue.enqueue(element);
+        queue.enqueue(element);

```

Comment:
Already `queueTriple` :smile:

Reply:
Fixed.

Reply:
Also changed the `Semaphore`s with negative initial values to `CountDownLatch`.